autoPyTorch/pipeline/components/training/trainer/base_trainer_choice.py

```diff
@@ -0,0 +1,479 @@
+            # can see this new network -- via an update, not overwrite of the pointer
+            state_dict = state_dict.result
+            X['network'].load_state_dict(state_dict)
+
+        # TODO: when have the optimizer code, the pynisher object might have failed
+        # We should process this function as Failure if so trough fit_function.exit_status
+        return cast(autoPyTorchComponent, self.choice)
+
+    def _fit(self, X: Dict[str, Any], y: Any = None, **kwargs: Any) -> torch.nn.Module:
+        """
+        Fits a component by using an input dictionary with pre-requisites
+
+        Args:
+            X (X: Dict[str, Any]): Dependencies needed by current component to perform fit
+            y (Any): not used. To comply with sklearn API
+
+        Returns:
+            A instance of self
+        """
+
+        # Comply with mypy
+        assert self.choice is not None
+
+        # Setup a Logger and other logging support
+        if 'use_tensorboard_logger' in X and X['use_tensorboard_logger']:
+            self.writer = SummaryWriter(log_dir=X['working_dir'])
+
+        if X["torch_num_threads"] > 0:
+            torch.set_num_threads(X["torch_num_threads"])
+
+        budget_tracker = BudgetTracker(
+            budget_type=X['budget_type'],
+            max_runtime=X['runtime'] if 'runtime' in X else None,
+            max_epochs=X['epochs'] if 'epochs' in X else None,
+        )
+
+        self.choice.prepare(
+            model=X['network'],
+            metrics=[m() for m in get_metrics(X['dataset_properties'])],
+            criterion=get_loss_instance(X['dataset_properties']),
+            budget_tracker=budget_tracker,
+            optimizer=X['optimizer'],
+            device=self.get_device(X),
+            logger=self.logger,
+            writer=self.writer,
+            metrics_during_training=X['metrics_during_training'],
+        )
+        total_parameter_count, trainable_parameter_count = self.count_parameters(X['network'])
+        self.run_summary = RunSummary(
+            total_parameter_count,
+            trainable_parameter_count,
+        )
+
+        epoch = 1
+
+        while True:
+
+            # prepare epoch
+            start_time = time.time()
+
+            self.choice.on_epoch_start(X=X, epoch=epoch)
+
+            # training
+            train_loss, train_metrics = self.choice.train_epoch(
+                train_loader=X['train_data_loader'],
+                epoch=epoch,
+            )
+
+            val_loss, val_metrics, test_loss, test_metrics = None, {}, None, {}
+            if self.eval_valid_each_epoch(X):
+                val_loss, val_metrics = self.choice.evaluate(X['val_data_loader'], epoch)
+                if 'test_data_loader' in X and X['test_data_loader']:
+                    test_loss, test_metrics = self.choice.evaluate(X['test_data_loader'], epoch)
+
+            # TODO: does it make sense to call all schedulers here?
```

Comment:
You could move the scheduler call to the (Base)Trainer, then we can account for calling them step-wise

Reply:
yes of course